packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -895,6 +1080,687 @@ void Add(
+          count[bucket]--;
+        }
+      }
+      whichBuf = 1 - whichBuf;
+      mask = mask << 4;
+      maskPos += 4;
+    }
+    if(whichBuf)
+    {
+      //need to deep copy from aux arrays to main
+      for(Ordinal i = lo; i < hi; i++)
+      {
+        CcolindOver(i) = CcolindAux(i);
+        CvalsOver(i) = CvalsAux(i);
+      }
+    }
+  }
+  
+  KOKKOS_INLINE_FUNCTION void operator()(const size_t i) const {
+    //3: Sort each row's colinds (permuting values at same time), then count unique colinds (write that to Crowptr(i))
+    //CrowptrTemp tells how many entries in each oversized row
+    Ordinal rowStart = CrowptrTemp(i);
+    Ordinal rowEnd = CrowptrTemp(i + 1);
+    Ordinal rowNum = rowEnd - rowStart;
+    radixSort2(rowStart, rowEnd);
+    //now that they are sorted, can sweep through and gather together values in same column
+    Ordinal finalPos = 0;
+    Ordinal overPos = 0;
+    while(overPos < rowNum)
+    {
+      //check if the two entries at overPos are the same col
+      if(CcolindOver(rowStart + overPos) == CcolindOver(rowStart + overPos + 1) && (overPos + 1 < rowNum))
+      {
+        //they are the same col, so combine the two entries at finalPos
+        //Then increment overPos an extra time since just processed two entries
+        CcolindOver(rowStart + finalPos) = CcolindOver(rowStart + overPos);
+        CvalsOver(rowStart + finalPos) = CvalsOver(rowStart + overPos) + CvalsOver(rowStart + overPos + 1);
+        overPos++;
+      }
+      else
+      {
+        //they are not the same col, so just copy the one entry at overPos to finalPos
+        CcolindOver(rowStart + finalPos) = CcolindOver(rowStart + overPos);
+        CvalsOver(rowStart + finalPos) = CvalsOver(rowStart + overPos);
+      }
+      overPos++;
+      finalPos++;
+    }
+    Crowcounts(i) = finalPos;
+  }
+  const RowPtrs& CrowptrTemp;
+  RowPtrs& Crowcounts;
+  ColInds& CcolindOver;
+  Values& CvalsOver;
+  //Two scratch arrays, used for sorting (with same size as CcolindOver and CvalsOver)
+  ColInds& CcolindAux;
+  Values& CvalsAux;
+  int numRadixSweeps;
+};
+
+template<typename Ordinal, typename Values, typename RowPtrs, typename ColInds>
+struct CondenseArraysFunctor
+{
+  CondenseArraysFunctor(const RowPtrs& Crowptr_, const RowPtrs& CrowptrTemp_, const ColInds& CcolindOver_, ColInds& Ccolind_, const Values& CvalsOver_, Values& Cvals_) :
+    Crowptr(Crowptr_),
+    CrowptrTemp(CrowptrTemp_),
+    CcolindOver(CcolindOver_),
+    Ccolind(Ccolind_),
+    CvalsOver(CvalsOver_),
+    Cvals(Cvals_)
+  {}
+  //4: Copy values from oversized Cvals/Ccolind to correctly sized
+  KOKKOS_INLINE_FUNCTION void operator()(const size_t i) const
+  {
+    Ordinal rowNum = Crowptr(i + 1) - Crowptr(i);
+    for(Ordinal j = 0; j < rowNum; j++)
+    {
+      Ccolind(Crowptr(i) + j) = CcolindOver(CrowptrTemp(i) + j);
+      Cvals(Crowptr(i) + j) = CvalsOver(CrowptrTemp(i) + j);
+    }
+  }
+  const RowPtrs& Crowptr;
```

Comment:
See above note on Kokkos::View needing to be stored by value in functors, not by reference.